core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlock.java

```diff
@@ -159,26 +159,38 @@ public Block copyPositions(int[] positions, int offset, int length)
     {
         checkArrayRange(positions, offset, length);
 
+        int[] newOffsets = new int[length + 1];
         int finalLength = 0;
-        for (int i = offset; i < offset + length; i++) {
-            finalLength += getSliceLength(positions[i]);
+        for (int i = 0; i < length; i++) {
+            int position = positions[offset + i];
+            finalLength += getSliceLength(position);
+            newOffsets[i + 1] = finalLength;
         }
+
         SliceOutput newSlice = Slices.allocate(finalLength).getOutput();
-        int[] newOffsets = new int[length + 1];
         boolean[] newValueIsNull = null;
-        if (valueIsNull != null) {
-            newValueIsNull = new boolean[length];
-        }
-
-        for (int i = 0; i < length; i++) {
-            int position = positions[offset + i];
-            if (!isEntryNull(position)) {
-                newSlice.writeBytes(slice, getPositionOffset(position), getSliceLength(position));
+        if (length > 0) {
+            if (valueIsNull != null) {
+                newValueIsNull = new boolean[length];
+                newValueIsNull[0] = valueIsNull[positions[offset] + arrayOffset];
             }
-            else if (newValueIsNull != null) {
-                newValueIsNull[i] = true;
+            int currentStart = getPositionOffset(positions[offset]);
+            int currentEnd = getPositionOffset(positions[offset] + 1);
+            for (int i = 1; i < length; i++) {
+                int position = positions[offset + i];
+                int currentOffset = getPositionOffset(position);
```

Comment:
I don't think offsets are guaranteed to make sense for null positions (add a test?)

Reply:
The bytes at a particular offset in slice for a null position can be junk, we don't care about that. However, the offset positions need to be valid in existing code as well, otherwise the result of `getSliceLength` for the next non-null position could go wrong.